File: HW_7/src/Board/BoardArray1D.java
package Board;
import java.io.*;

/**
 * Extended concrete class which represents whole class.
 * Its implementation is based on two one dimentional array,
 * @see java.lang.Cloneable
 * @see Board.AbstractBoard
 * */
public class BoardArray1D extends AbstractBoard implements Cloneable {

    private int puzzle_map[];

    /**
     * No parameter constructor.
     * Default x size = 3.
     * Default y size = 3.
     * */
    public BoardArray1D( ) {
        this( DEFAULT_X_SIZE , DEFAULT_Y_SIZE );
    }


    /**
     * Copy Constructor.
     * Create a new board which is equivalent the given as a parameter.
     * @param other Using deep copy method , copy the parameter to current board.
     * Defaulted x size = 3.
     * Defaulted y size = 3
     */
    public BoardArray1D( BoardArray1D other ) {
        super();

        setSize( other.get_xSize() , other.get_ySize() );

        for(int j = 0 ; j < this.get_ySize() ; ++j)
            for(int i = 0 ; i < this.get_xSize() ; ++i)
                puzzle_map[ j * this.get_xSize() + i ] = other.puzzle_map[ j * this.get_xSize() + i ];
    }

    /**
     * Create a new board which is equivalent the configuration
     * that is created according to given parameter.
     * @param xSize x size of the board , defaulted x size = 3.
     * @param ySize y size of the board , defaulted y size = 3.
     */
    public BoardArray1D( int xSize , int ySize ) {
        super();

        puzzle_map = null;
        setSize( xSize , ySize );

    }


    /**
     * Create a new board according to given file by the constructor.
     * Defaulted x size = 3.
     * Defaulted y size = 3.
     * @param filename File name which is going to be opened.
     * @throws IOException if file not exist or not opened.
     */
    public BoardArray1D( String filename) throws IOException{
        super();

        puzzle_map = null;
        try
        {
            readFromFile(filename);
        }
        catch ( IOException  ExceptionPocket ) {
            System.out.println( ExceptionPocket.getMessage() );
        }
    }


    /**
     * Set the board with given size.
     * @param xSizeOut x size of the board.
     * @param ySizeOut y size of the board.
     */
    @Override
    public void setSize( int xSizeOut , int ySizeOut ) {

        if( puzzle_map != null) {
            puzzle_map = null;
        }

        /* Base class virtual call */
        super.setSize( xSizeOut , ySizeOut );

        puzzle_map = new int [ xSizeOut * ySizeOut ];

        for(int j = 0 ; j < get_ySize() ; ++j)
            for(int i = 0 ; i < get_xSize() ; ++i)
                puzzle_map[ j * get_xSize() + i ] = -1;

        /* Whatever is goint to be built */

        reset();
    }

    /**
     * Obtaining specific cell content located in board.
     * @param x x coordinate of the number.
     * @param y y coordinate of the number.
     * @return Return the number that is located in x and y.
     */
    @Override
    public int cell( int x , int y ) {

        if( (x < 0 || x > (get_xSize() - 1) ) || (y < 0 || y > (get_ySize() - 1) ) )
        {
            System.err.println("Invalid argument ! Error , operator() ");
            System.exit(-1);
        }

        return puzzle_map[ y * get_xSize() + x ];
    }

    /**
     * Replace the given coordinate number with new.
     * @param x x coordinate of the number.
     * @param y y coordinate of the number.
     * @param content content which is going to be replaced with previous.
     * These method are implemented in extended classes.
     */
    @Override
    protected void cell( int x , int y , int content ) {

        if( (x < 0 || x > ( get_xSize() - 1 ) ) || (y < 0 || y > ( get_ySize() - 1 ) ) )
        {
            System.err.println("Invalid argument ! Error , operator() ");
            System.exit(-1);
        }
        puzzle_map[ y * get_xSize() + x ] = content;

    }

    /**
     * Return a deep copied new BoardArray1D that is equivalent to current.
     * @throws CloneNotSupportedException if Error during the process.
     */
    @Override
    public Object clone() throws CloneNotSupportedException{
        return new BoardArray1D(this);
    }
}

File: HW_5/src/BoardArray1D.cpp
#include <iostream>
#include "BoardArray1D.h"
#include "AbstractBoard.h"

namespace NPuzzle
{
    using std::string;
    using std::cout;
    using std::cin;
    using std::cerr;
    using std::endl;


    BoardArray1D::BoardArray1D(int xSize, int ySize) : AbstractBoard()
        , puzzle_map(nullptr)
    {
        setSize( xSize , ySize );

        /* Intentionally left blank */
    }

    BoardArray1D::BoardArray1D( const BoardArray1D& other) : AbstractBoard( other )
        , puzzle_map(nullptr)
    {
        inc_Number_Of_Boards();
        
        setSize( get_xSize() , get_ySize() );

        for(int j = 0 ; j < get_ySize() ; ++j)
            for(int i = 0 ; i < get_xSize() ; ++i)
                puzzle_map[ j * get_xSize() + i ] = other.puzzle_map[ j * get_xSize() + i ];
    }

    BoardArray1D::BoardArray1D( const string& filename) : AbstractBoard()
        , puzzle_map(nullptr)
    {
        readFromFile( filename );
    }


    void BoardArray1D::setSize( const int& xSize , const int& ySize )
    {
        if( puzzle_map != nullptr)
        {
            delete [] puzzle_map;
            puzzle_map = nullptr;
        }

        /* Base class virtual call */
        AbstractBoard::setSize( xSize , ySize );
        puzzle_map = new int [ xSize * ySize ];

        for(int j = 0 ; j < get_ySize() ; ++j)
            for(int i = 0 ; i < get_xSize() ; ++i)
                puzzle_map[ j * get_xSize() + i ] = -1;

        /* Whatever is goint to be built */
        
        reset();
    }

    int& BoardArray1D::operator()(int x , int y)
    {
        if( (x < 0 || x > (get_xSize() - 1) ) || (y < 0 || y > (get_ySize() - 1) ) )
        {
            cerr << endl << "Invalid argument ! Error , operator() " << endl << endl;
            exit(1);
        }

        return puzzle_map[ y * get_xSize() + x ];
    }

    const int& BoardArray1D::operator()(int x , int y) const
    {
        if( (x < 0 || x > (get_xSize() - 1) ) || (y < 0 || y > (get_ySize() - 1) ) )
        {
            cerr << endl << "Invalid argument ! Error , operator() " << endl << endl;
            exit(1);
        }
        
        return puzzle_map[ y * get_xSize() + x ];
    }

    BoardArray1D& BoardArray1D::operator=(const BoardArray1D& other)
    {
        if( puzzle_map != nullptr )
        {
            delete [] puzzle_map;
            puzzle_map = nullptr;
        }

        AbstractBoard::operator=( other );

        puzzle_map = new int [ get_xSize() * get_ySize() ];
        
        for(int j = 0 ; j < get_ySize() ; ++j)
            for(int i = 0 ; i< get_xSize() ; ++i)
                puzzle_map[ j * get_xSize() + i ] = other.puzzle_map[ j * get_xSize() + i ];

        return *this;
    }

    BoardArray1D::~BoardArray1D()
    {
        delete [] puzzle_map;
        puzzle_map = nullptr;
    }
}